### PR TITLE
CI: run tests in parallel (pytest-xdist)

### LINKED ISSUE #3263: test_sparse doesn't work with pytest-xdist
`pytest -n 4 test_sparse.py` results in

```
====================================================================================== ERRORS ======================================================================================
_______________________________________________________________________________ ERROR collecting gw0 _______________________________________________________________________________
Different tests were collected between gw1 and gw0. The difference is:
--- gw1

+++ gw0

@@ -11,7 +11,7 @@

 xarray/tests/test_sparse.py::test_variable_method[obj.any(*(), **{})-False]
 xarray/tests/test_sparse.py::test_variable_method[obj.astype(*(), **{'dtype': <class 'int'>})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.clip(*(), **{'min': 0, 'max': 1})-True]
-xarray/tests/test_sparse.py::test_variable_method[obj.coarsen(*(), **{'windows': {'x': 2}, 'func': <function sum at 0x7fc7303f7d08>})-True]
+xarray/tests/test_sparse.py::test_variable_method[obj.coarsen(*(), **{'windows': {'x': 2}, 'func': <function sum at 0x7f6009fa4d08>})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.compute(*(), **{})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.conj(*(), **{})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.copy(*(), **{})-True]
@@ -49,7 +49,7 @@

 xarray/tests/test_sparse.py::test_variable_method[obj.prod(*(), **{})-False]
 xarray/tests/test_sparse.py::test_variable_method[obj.quantile(*(), **{'q': 0.5})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.rank(*(), **{'dim': 'x'})-False]
-xarray/tests/test_sparse.py::test_variable_method[obj.reduce(*(), **{'func': <function sum at 0x7fc7303f7d08>, 'dim': 'x'})-True]
+xarray/tests/test_sparse.py::test_variable_method[obj.reduce(*(), **{'func': <function sum at 0x7f6009fa4d08>, 'dim': 'x'})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.rolling_window(*(), **{'dim': 'x', 'window': 2, 'window_dim': 'x_win'})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.shift(*(), **{'x': 2})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.std(*(), **{})-False]
@@ -144,11 +144,11 @@

 xarray/tests/test_sparse.py::test_dataarray_method[obj.median(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.min(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.notnull(*(), **{})-False]
-xarray/tests/test_sparse.py::test_dataarray_method[obj.pipe(*(<function sum at 0x7fc7303f7d08>,), **{'axis': 1})-True]
+xarray/tests/test_sparse.py::test_dataarray_method[obj.pipe(*(<function sum at 0x7f6009fa4d08>,), **{'axis': 1})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.prod(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.quantile(*(), **{'q': 0.5})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.rank(*('x',), **{})-False]
-xarray/tests/test_sparse.py::test_dataarray_method[obj.reduce(*(<function sum at 0x7fc7303f7d08>,), **{'dim': 'x'})-False]
+xarray/tests/test_sparse.py::test_dataarray_method[obj.reduce(*(<function sum at 0x7f6009fa4d08>,), **{'dim': 'x'})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.reindex_like(*(<xarray.DataArray 'test' (x: 10, y: 5)>\n<COO: shape=(10, 5), dtype=float64, nnz=5, fill_value=0.0>\nCoordinates:\n  * x        (x) float64 0.5 1.5 2.5 3.5 4.5 5.5 6.5 7.5 8.5 9.5\n  * y        (y) float64 0.5 1.5 2.5 3.5 4.5,), **{})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.roll(*(), **{'x': 2})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.sel(*(), **{'x': [0, 1, 2], 'y': [2, 3]})-True]
_______________________________________________________________________________ ERROR collecting gw2 _______________________________________________________________________________
Different tests were collected between gw1 and gw2. The difference is:
--- gw1

+++ gw2

@@ -11,7 +11,7 @@

 xarray/tests/test_sparse.py::test_variable_method[obj.any(*(), **{})-False]
 xarray/tests/test_sparse.py::test_variable_method[obj.astype(*(), **{'dtype': <class 'int'>})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.clip(*(), **{'min': 0, 'max': 1})-True]
-xarray/tests/test_sparse.py::test_variable_method[obj.coarsen(*(), **{'windows': {'x': 2}, 'func': <function sum at 0x7fc7303f7d08>})-True]
+xarray/tests/test_sparse.py::test_variable_method[obj.coarsen(*(), **{'windows': {'x': 2}, 'func': <function sum at 0x7f657c314d08>})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.compute(*(), **{})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.conj(*(), **{})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.copy(*(), **{})-True]
@@ -49,7 +49,7 @@

 xarray/tests/test_sparse.py::test_variable_method[obj.prod(*(), **{})-False]
 xarray/tests/test_sparse.py::test_variable_method[obj.quantile(*(), **{'q': 0.5})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.rank(*(), **{'dim': 'x'})-False]
-xarray/tests/test_sparse.py::test_variable_method[obj.reduce(*(), **{'func': <function sum at 0x7fc7303f7d08>, 'dim': 'x'})-True]
+xarray/tests/test_sparse.py::test_variable_method[obj.reduce(*(), **{'func': <function sum at 0x7f657c314d08>, 'dim': 'x'})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.rolling_window(*(), **{'dim': 'x', 'window': 2, 'window_dim': 'x_win'})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.shift(*(), **{'x': 2})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.std(*(), **{})-False]
@@ -118,7 +118,7 @@

 xarray/tests/test_sparse.py::test_dataarray_method[obj.sel(*(), **{'x': [0, 1, 2]})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.shift(*(), **{})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.sortby(*('x',), **{'ascending': False})-True]
-xarray/tests/test_sparse.py::test_dataarray_method[obj.stack(*(), **{'z': {'y', 'x'}})-True]
+xarray/tests/test_sparse.py::test_dataarray_method[obj.stack(*(), **{'z': {'x', 'y'}})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.transpose(*(), **{})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.broadcast_equals(*(<xarray.Variable (x: 10, y: 5)>\n<COO: shape=(10, 5), dtype=float64, nnz=5, fill_value=0.0>,), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.equals(*(<xarray.Variable (x: 10, y: 5)>\n<COO: shape=(10, 5), dtype=float64, nnz=5, fill_value=0.0>,), **{})-False]
@@ -144,11 +144,11 @@

 xarray/tests/test_sparse.py::test_dataarray_method[obj.median(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.min(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.notnull(*(), **{})-False]
-xarray/tests/test_sparse.py::test_dataarray_method[obj.pipe(*(<function sum at 0x7fc7303f7d08>,), **{'axis': 1})-True]
+xarray/tests/test_sparse.py::test_dataarray_method[obj.pipe(*(<function sum at 0x7f657c314d08>,), **{'axis': 1})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.prod(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.quantile(*(), **{'q': 0.5})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.rank(*('x',), **{})-False]
-xarray/tests/test_sparse.py::test_dataarray_method[obj.reduce(*(<function sum at 0x7fc7303f7d08>,), **{'dim': 'x'})-False]
+xarray/tests/test_sparse.py::test_dataarray_method[obj.reduce(*(<function sum at 0x7f657c314d08>,), **{'dim': 'x'})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.reindex_like(*(<xarray.DataArray 'test' (x: 10, y: 5)>\n<COO: shape=(10, 5), dtype=float64, nnz=5, fill_value=0.0>\nCoordinates:\n  * x        (x) float64 0.5 1.5 2.5 3.5 4.5 5.5 6.5 7.5 8.5 9.5\n  * y        (y) float64 0.5 1.5 2.5 3.5 4.5,), **{})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.roll(*(), **{'x': 2})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.sel(*(), **{'x': [0, 1, 2], 'y': [2, 3]})-True]
_______________________________________________________________________________ ERROR collecting gw3 _______________________________________________________________________________
Different tests were collected between gw1 and gw3. The difference is:
--- gw1

+++ gw3

@@ -11,7 +11,7 @@

 xarray/tests/test_sparse.py::test_variable_method[obj.any(*(), **{})-False]
 xarray/tests/test_sparse.py::test_variable_method[obj.astype(*(), **{'dtype': <class 'int'>})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.clip(*(), **{'min': 0, 'max': 1})-True]
-xarray/tests/test_sparse.py::test_variable_method[obj.coarsen(*(), **{'windows': {'x': 2}, 'func': <function sum at 0x7fc7303f7d08>})-True]
+xarray/tests/test_sparse.py::test_variable_method[obj.coarsen(*(), **{'windows': {'x': 2}, 'func': <function sum at 0x7f6f284e3d08>})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.compute(*(), **{})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.conj(*(), **{})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.copy(*(), **{})-True]
@@ -49,7 +49,7 @@

 xarray/tests/test_sparse.py::test_variable_method[obj.prod(*(), **{})-False]
 xarray/tests/test_sparse.py::test_variable_method[obj.quantile(*(), **{'q': 0.5})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.rank(*(), **{'dim': 'x'})-False]
-xarray/tests/test_sparse.py::test_variable_method[obj.reduce(*(), **{'func': <function sum at 0x7fc7303f7d08>, 'dim': 'x'})-True]
+xarray/tests/test_sparse.py::test_variable_method[obj.reduce(*(), **{'func': <function sum at 0x7f6f284e3d08>, 'dim': 'x'})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.rolling_window(*(), **{'dim': 'x', 'window': 2, 'window_dim': 'x_win'})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.shift(*(), **{'x': 2})-True]
 xarray/tests/test_sparse.py::test_variable_method[obj.std(*(), **{})-False]
@@ -144,11 +144,11 @@

 xarray/tests/test_sparse.py::test_dataarray_method[obj.median(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.min(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.notnull(*(), **{})-False]
-xarray/tests/test_sparse.py::test_dataarray_method[obj.pipe(*(<function sum at 0x7fc7303f7d08>,), **{'axis': 1})-True]
+xarray/tests/test_sparse.py::test_dataarray_method[obj.pipe(*(<function sum at 0x7f6f284e3d08>,), **{'axis': 1})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.prod(*(), **{})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.quantile(*(), **{'q': 0.5})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.rank(*('x',), **{})-False]
-xarray/tests/test_sparse.py::test_dataarray_method[obj.reduce(*(<function sum at 0x7fc7303f7d08>,), **{'dim': 'x'})-False]
+xarray/tests/test_sparse.py::test_dataarray_method[obj.reduce(*(<function sum at 0x7f6f284e3d08>,), **{'dim': 'x'})-False]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.reindex_like(*(<xarray.DataArray 'test' (x: 10, y: 5)>\n<COO: shape=(10, 5), dtype=float64, nnz=5, fill_value=0.0>\nCoordinates:\n  * x        (x) float64 0.5 1.5 2.5 3.5 4.5 5.5 6.5 7.5 8.5 9.5\n  * y        (y) float64 0.5 1.5 2.5 3.5 4.5,), **{})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.roll(*(), **{'x': 2})-True]
 xarray/tests/test_sparse.py::test_dataarray_method[obj.sel(*(), **{'x': [0, 1, 2], 'y': [2, 3]})-True]
================================================================================= warnings summary =================================================================================
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/client.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/client.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/client.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/client.py:2
  /home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/client.py:2: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import defaultdict, Iterator

/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/publish.py:1
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/publish.py:1
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/publish.py:1
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/publish.py:1
  /home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/publish.py:1: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import MutableMapping

/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
/home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2
  /home/deepak/miniconda3/envs/dcpy/lib/python3.7/site-packages/distributed/scheduler.py:2: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import defaultdict, deque, OrderedDict, Mapping, Set

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=======================================================================
```

### PR DESCRIPTION
**Pull request by mathause**
- [x] Closes #3263, potentially part of #4671

It's worth a try.

### CODE REVIEW

File: xarray/tests/test_sparse.py [hunk 1]
```diff
@@ -359,7 +365,7 @@ def test_dataarray_property(prop):
         (do("sel", x=[0, 1, 2]), True),
         (do("shift"), True),
         (do("sortby", "x", ascending=False), True),
-        (do("stack", z={"x", "y"}), True),
+        (do("stack", z=sorted({"x", "y"})), True),  # sorted for pytest-xdist
```
Comment (dcherian) [inline xarray/tests/test_sparse.py:368]:
```suggestion
        (do("stack", z=["x", "y"], True),  # sorted for pytest-xdist
```

should work?

File: xarray/tests/test_sparse.py [hunk 2]
```diff
@@ -62,7 +62,13 @@ def __init__(self, meth, *args, **kwargs):
         self.kwargs = kwargs
 
     def __call__(self, obj):
-        return getattr(obj, self.meth)(*self.args, **self.kwargs)
+
+        # cannot pass np.sum when using pytest-xdist
```
Comment (dcherian) [inline xarray/tests/test_sparse.py:66]: nice find!

**Review by dcherian — APPROVED** [review]
Great fix even if it only works locally

File: xarray/tests/test_sparse.py [hunk 3]
```diff
@@ -359,7 +365,7 @@ def test_dataarray_property(prop):
         (do("sel", x=[0, 1, 2]), True),
         (do("shift"), True),
         (do("sortby", "x", ascending=False), True),
-        (do("stack", z={"x", "y"}), True),
+        (do("stack", z=["x", "y"], True),
```
Comment (dcherian) [inline xarray/tests/test_sparse.py:368]:
```suggestion
        (do("stack", z=["x", "y"]), True),
```

File: ci/azure/unit-tests.yml [hunk 4]
```diff
@@ -11,7 +11,7 @@ steps:
 # https://github.com/microsoft/azure-pipelines-tasks/issues/9302
 - bash: |
     source activate xarray-tests
-    $(environment_variables) pytest \
+    $(environment_variables) -n 2 pytest \
```
Comment (mathause) [inline ci/azure/unit-tests.yml:14]:
```suggestion
    $(environment_variables) pytest -n 2 \
```

**Comment by dcherian** [comment]
Wow! Looks like a 30-40% improvement

 (I'm not sure why the latest commit triggered two runs, but one of them has finished: https://dev.azure.com/xarray/xarray/_build?definitionId=1&_a=summary)

### Before:
image


### After 
image

**Comment by mathause** [comment]
> Looks like a 30-40% improvement

Nice! I think we get 2 hypter-threaded cores. So I can also try `-n 4` and see if that helps. I wonder if there is any downside to this?

**Comment by andersy005** [comment]
> Nice! I think we get 2 hypter-threaded cores. So I can also try -n 4 and see if that helps. I wonder if there is any downside to this?

In the past, I've run into race conditions when using pytest-xdist for tests that required I/O tasks such as file downloading. So, it's likely everything will run fine with `-n 4` provided we don't have tests with tasks that are likely to create race conditions.

**Comment by mathause** [comment]
> In the past, I've run into race conditions when using pytest-xdist for tests that required I/O tasks such as file downloading

It's not too difficult to remove the `-n 4` and pytest-xdist again if we run into problems. It may, however, be more difficult to find out that random failures are due to race conditions...

**Review by andersy005 — APPROVED** [review]
Looks great... :shipit: ship it

**Comment by dcherian** [comment]
I updated my previous comment. It looks like -n 4 saved us 5 minutes on windows and didn't affect the others.

**Comment by dcherian** [comment]
Great. Thanks @mathause this is a great improvement!

**Comment by dcherian** [comment]
sorry! didn't realize you were still working on it.

**Comment by mathause** [comment]
If it passes I am finished ;-)

**Pull request merged**